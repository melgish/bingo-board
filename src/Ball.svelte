<script>
  import { createEventDispatcher } from 'svelte'
  // true when highlighted
  export let lit = false
  const dispatch = createEventDispatcher()

  $: hot = false

  function flip() {
    hot = true
    dispatch('flip')
    setTimeout(() => (hot = false), 2000)
  }
</script>

<div role="button" class:hot class:lit on:click={flip}>
  <span>
    <slot />
  </span>
</div>

<style>
  div {
    display: grid;
    justify-items: center;
    align-content: center;
    text-align: center;
  }

  span {
    display: grid;
    width: 40pt;
    height: 36pt;
    border-radius: 50px;
    font-size: 24pt;
    font-weight: bold;
  }
  @media screen {
    div {
      background-color: black;
      color: white;
      transition: background-color 2s ease;
      border: 1px solid #300;
    }
    div.hot {
      background-color: #cc0;
    }
    span {
      border-radius: 50%;
      border: 1pt solid currentColor;
      transition: background-color 1s ease;
    }
    .lit span {
      background-color: white;
      color: black;
    }
  }

  @media print {
    div {
      background-color: white;
      color: black;
      border: 1px solid black;
    }
    span {
      border: 1px solid silver;
    }
  }
</style>
