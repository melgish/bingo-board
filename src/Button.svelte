<script>
  export let disabled = false
</script>

<button on:click type="button" {disabled}>
  <slot />
</button>

<style>
  button {
    font-weight: bold;
    background-color: #600;
    border-color: #300;
    color: white;
    padding: 0.5rem 1rem;
    border-radius: 0.5rem;
  }
  button:disabled {
    background-color: #66000080;
  }
</style>
