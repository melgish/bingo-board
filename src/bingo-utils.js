import gen from "random-seed"
import shuffle from 'shuffle-array';

export const GAME = 'COVID';

// Array of indexes 0 - 4 because I'm that lazy
const BINGO = GAME.split('').map((v, k) => k);
const FREE = 2;
const RNG = gen.create();

// Array of all available bingo numbers arranged in board layout
// [1, 2, ... 15]
// [16, 17, ... 30]
// ...
export const BOARD = BINGO.map(
  i => Array.from({length: 15}, (v, k) => (i * 15) + k + 1)
);


/**
 * Create new 5 digit seed
 *
 * @return {number} random 5 digit number
 */
export function getSeed() {
  return RNG.intBetween(10000, 100000);
}

/**
 * Generate and return card for given seed
 *
 * @param {number} seed
 */
export function getCard(seed) {
  const picks = 5;
  // Use repeatable random number generator.
  const rng = gen.create(seed).random;
  // Capture 5 random values from each letter
  const cols = BINGO.map(i => shuffle.pick(BOARD[i], { picks, rng }));
  // Clear the free space
  cols[FREE][FREE] = 0;
  // Arrange in rows
  const rows = BINGO.map(i => BINGO.map(j => cols[j][i]));

  return { seed, rows };
}

