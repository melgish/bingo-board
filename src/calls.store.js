import marks from './marks.store';
// calls is a singleton instance of the marks store.
export default marks();
