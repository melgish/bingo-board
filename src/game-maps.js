export const coverAll = [
  [1, 1, 1, 1, 1],
  [1, 1, 1, 1, 1],
  [1, 1, 1, 1, 1],
  [1, 1, 1, 1, 1],
  [1, 1, 1, 1, 1],
]

export const bigBox = [
  [1, 1, 1, 1, 1],
  [1, 0, 0, 0, 1],
  [1, 0, 0, 0, 1],
  [1, 0, 0, 0, 1],
  [1, 1, 1, 1, 1],
]

export const smallBox = [
  [0, 0, 0, 0, 0],
  [0, 1, 1, 1, 0],
  [0, 1, 0, 1, 0],
  [0, 1, 1, 1, 0],
  [0, 0, 0, 0, 0],
]

export const bigCorners = [
  [1, 1, 0, 1, 1],
  [1, 1, 0, 1, 1],
  [0, 0, 0, 0, 0],
  [1, 1, 0, 1, 1],
  [1, 1, 0, 1, 1],
]

export const letterX = [
  [1, 0, 0, 0, 1],
  [0, 1, 0, 1, 0],
  [0, 0, 1, 0, 0],
  [0, 1, 0, 1, 0],
  [1, 0, 0, 0, 1],
]

export const letterT = [
  [1, 1, 1, 1, 1],
  [0, 0, 1, 0, 0],
  [0, 0, 1, 0, 0],
  [0, 0, 1, 0, 0],
  [0, 0, 1, 0, 0],
]

export const cross = [
  [0, 0, 1, 0, 0],
  [0, 0, 1, 0, 0],
  [1, 1, 1, 1, 1],
  [0, 0, 1, 0, 0],
  [0, 0, 1, 0, 0],
]

export const letterL = [
  [1, 0, 0, 0, 0],
  [1, 0, 0, 0, 0],
  [1, 0, 0, 0, 0],
  [1, 0, 0, 0, 0],
  [1, 1, 1, 1, 1],
]

export const normals = [
  [
    [1, 1, 1, 1, 1],
    [0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0],
  ],
  [
    [0, 0, 0, 0, 0],
    [1, 1, 1, 1, 1],
    [0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0],
  ],
  [
    [0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0],
    [1, 1, 1, 1, 1],
    [0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0],
  ],
  [
    [0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0],
    [1, 1, 1, 1, 1],
    [0, 0, 0, 0, 0],
  ],
  [
    [0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0],
    [1, 1, 1, 1, 1],
  ],
  [
    [1, 0, 0, 0, 0],
    [1, 0, 0, 0, 0],
    [1, 0, 0, 0, 0],
    [1, 0, 0, 0, 0],
    [1, 0, 0, 0, 0],
  ],
  [
    [0, 1, 0, 0, 0],
    [0, 1, 0, 0, 0],
    [0, 1, 0, 0, 0],
    [0, 1, 0, 0, 0],
    [0, 1, 0, 0, 0],
  ],
  [
    [0, 0, 1, 0, 0],
    [0, 0, 1, 0, 0],
    [0, 0, 1, 0, 0],
    [0, 0, 1, 0, 0],
    [0, 0, 1, 0, 0],
  ],
  [
    [0, 0, 0, 1, 0],
    [0, 0, 0, 1, 0],
    [0, 0, 0, 1, 0],
    [0, 0, 0, 1, 0],
    [0, 0, 0, 1, 0],
  ],
  [
    [0, 0, 0, 0, 1],
    [0, 0, 0, 0, 1],
    [0, 0, 0, 0, 1],
    [0, 0, 0, 0, 1],
    [0, 0, 0, 0, 1],
  ],
  [
    [1, 0, 0, 0, 0],
    [0, 1, 0, 0, 0],
    [0, 0, 1, 0, 0],
    [0, 0, 0, 1, 0],
    [0, 0, 0, 0, 1],
  ],
  [
    [0, 0, 0, 0, 1],
    [0, 0, 0, 1, 0],
    [0, 0, 1, 0, 0],
    [0, 1, 0, 0, 0],
    [1, 0, 0, 0, 0],
  ],
]

export const GAME_MAPS = [
  bigBox,
  smallBox,
  bigCorners,
  letterX,
  letterT,
  cross,
  letterL,
  coverAll,
]
