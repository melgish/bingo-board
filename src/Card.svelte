<script>
  import Ball from "./Ball.svelte";
  import Letter from "./Letter.svelte";
  import { GAME } from './bingo-utils.js';

  export let card = { rows: [], seed: 0 };
  export let calls = {};
</script>

<style>
  .card {
    display: grid;
    grid-template-columns: repeat(5, 1fr);
    grid-template-rows: repeat(5, 3rem);
    user-select: none;
    margin-bottom: 1rem;
    justify-items: stretch;
    align-content: center;
    background-color: silver;
    border: 2px solid silver;
    grid-gap: 0;
    color: black;
  }

  .card .foot {
    grid-column-start: 1;
    grid-column-end: 6;
    text-align: center;
    padding: 0.25rem 0;
  }

  .foot {
    font-weight: bold;
  }

  @media print {
    .card {
      color: black;
      border: 1px solid black;
    }
  }
</style>

<div class="card">
  {#each GAME as letter}
    <Letter {letter} />
  {/each}

  {#each card.rows as row}
    {#each row as col}
      <Ball lit={calls[col]}>{col || 'free'}</Ball>
    {/each}
  {/each}
  <div class="foot">{card.seed}</div>
</div>
