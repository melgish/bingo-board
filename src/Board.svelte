<script>
  import Button from "./Button.svelte";
  import Ball from "./Ball.svelte";
  import Letter from "./Letter.svelte";
  import Card from "./Card.svelte";
  import GameMaps from "./GameMaps.svelte";
  import calls from "./calls.store.js";
  import { getCard, BOARD, GAME } from "./bingo-utils.js";

  /**
   * card being inspected
   */
  let card = null;

  /**
   * Seed for input / card being inspected
   */
  let seed = null;

  /**
   *  Create rows to display letter
   */
  let rows = GAME.split("").map((letter, i) => ({
    letter,
    balls: BOARD[i]
  }));

  /**
   * Clear the card being inspected
   */
  function clearCard() {
    card = null;
    seed = null;
  }

  /**
   * Check the card made from current seed
   */
  function checkCard() {
    if (seed > 9999) {
      card = getCard(seed);
    }
  }

  /**
   * Reset the game
   */
  function resetGame() {
    calls.reset();
    clearCard();
  }

  /**
   * Flip a single ball
   *
   * @param {number} ball ball to flip.
   */
  function flip(ball) {
    calls.flip(ball);
  }
</script>

<style>
  .board {
    display: grid;
    grid-template-columns: repeat(16, 1fr);
    grid-template-rows: repeat(5, 3rem);
    grid-gap: 0.125rem;
    user-select: none;
    margin-bottom: .25rem;
    background-color: #333;
    justify-items: stretch;
    align-content: center;
    padding: 2px;
    border: 1px solid #600;
  }
  input {
    font-size: 12pt;
    font-weight: bold;
    color: #600;
    border-radius: 0.5rem;
    border-color: #600;
  }
  .check-card {
    width: 50%;
  }

 </style>

<div class="board">
  {#each rows as row}
    <Letter letter={row.letter} />
    {#each row.balls as ball}
      <Ball lit={$calls[ball]} on:flip={() => flip(ball)}>{ball}</Ball>
    {/each}
  {/each}
</div>

<GameMaps></GameMaps>
<div>
  <Button on:click={resetGame}>Reset Game</Button>
  <input
    type="number"
    min="10000"
    max="99999"
    placeholder="card #"
    bind:value={seed}
    disabled={card} />
  {#if card}
    <Button on:click={clearCard}>Clear</Button>
  {:else}
    <Button on:click={checkCard} disabled={!(seed && seed > 9999)}>
      Check Card
    </Button>
  {/if}
</div>
{#if card}
<div class="check-card">
  <Card {card} calls={$calls} />
</div>
{/if}
