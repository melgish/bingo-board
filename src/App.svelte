<script>
  import { Router, Route, Link } from "svelte-routing"
  import { GAME } from "./bingo-utils.js"
  import Board from "./Board.svelte"
  import Cards from "./Cards.svelte"

  export let url = ""
</script>

<div class="contianer">
  <Router {url}>
    <nav class="no-print">
      <div class="nav-logo">Nick's Cheapo {GAME} Game:</div>
      <div class="nav-item">
        <Link to="/">Board</Link>
      </div>
      <div class="nav-item">
        <Link to="/cards">Cards</Link>
      </div>
    </nav>
    <div>
      <Route path="cards"><Cards /></Route>
      <Route><Board /></Route>
    </div>
  </Router>
</div>

<style>
  .contianer {
    padding: 1rem 2rem;
  }
  nav {
    margin-bottom: 4px;
  }
  .nav-logo {
    display: inline-block;
    font-size: 16pt;
    font-weight: bold;
    font-variant: small-caps;
    color: #600;
  }
  .nav-item {
    border: 2px solid silver;
    border-radius: 1rem;
    padding: 0.125rem 0.75rem;
    font-size: 14pt;
    display: inline-block;
    margin-top: -4px;
  }
</style>
