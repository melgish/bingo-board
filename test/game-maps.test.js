import { GAME_MAPS } from '../src/game-maps'

describe('GAME_MAPS', () => {
  it('should represent games', () => {
    expect(GAME_MAPS).toMatchSnapshot()
  })
})
