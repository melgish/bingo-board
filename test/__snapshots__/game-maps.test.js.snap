// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GAME_MAPS should represent games 1`] = `
Array [
  Array [
    Array [
      1,
      1,
      1,
      1,
      1,
    ],
    Array [
      1,
      0,
      0,
      0,
      1,
    ],
    Array [
      1,
      0,
      0,
      0,
      1,
    ],
    Array [
      1,
      0,
      0,
      0,
      1,
    ],
    Array [
      1,
      1,
      1,
      1,
      1,
    ],
  ],
  Array [
    Array [
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      0,
      1,
      1,
      1,
      0,
    ],
    Array [
      0,
      1,
      0,
      1,
      0,
    ],
    Array [
      0,
      1,
      1,
      1,
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
    ],
  ],
  Array [
    Array [
      1,
      1,
      0,
      1,
      1,
    ],
    Array [
      1,
      1,
      0,
      1,
      1,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
    ],
    Array [
      1,
      1,
      0,
      1,
      1,
    ],
    Array [
      1,
      1,
      0,
      1,
      1,
    ],
  ],
  Array [
    Array [
      1,
      0,
      0,
      0,
      1,
    ],
    Array [
      0,
      1,
      0,
      1,
      0,
    ],
    Array [
      0,
      0,
      1,
      0,
      0,
    ],
    Array [
      0,
      1,
      0,
      1,
      0,
    ],
    Array [
      1,
      0,
      0,
      0,
      1,
    ],
  ],
  Array [
    Array [
      1,
      1,
      1,
      1,
      1,
    ],
    Array [
      0,
      0,
      1,
      0,
      0,
    ],
    Array [
      0,
      0,
      1,
      0,
      0,
    ],
    Array [
      0,
      0,
      1,
      0,
      0,
    ],
    Array [
      0,
      0,
      1,
      0,
      0,
    ],
  ],
  Array [
    Array [
      0,
      0,
      1,
      0,
      0,
    ],
    Array [
      0,
      0,
      1,
      0,
      0,
    ],
    Array [
      1,
      1,
      1,
      1,
      1,
    ],
    Array [
      0,
      0,
      1,
      0,
      0,
    ],
    Array [
      0,
      0,
      1,
      0,
      0,
    ],
  ],
  Array [
    Array [
      1,
      0,
      0,
      0,
      0,
    ],
    Array [
      1,
      0,
      0,
      0,
      0,
    ],
    Array [
      1,
      0,
      0,
      0,
      0,
    ],
    Array [
      1,
      0,
      0,
      0,
      0,
    ],
    Array [
      1,
      1,
      1,
      1,
      1,
    ],
  ],
  Array [
    Array [
      1,
      1,
      1,
      1,
      1,
    ],
    Array [
      1,
      1,
      1,
      1,
      1,
    ],
    Array [
      1,
      1,
      1,
      1,
      1,
    ],
    Array [
      1,
      1,
      1,
      1,
      1,
    ],
    Array [
      1,
      1,
      1,
      1,
      1,
    ],
  ],
]
`;
